hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryRunner.java

```diff
@@ -135,19 +141,29 @@ public Result runIndexOrPartitionScanQueryOnOwnedPartitions(Query query, boolean
         Iterable<QueryableEntry> entries = runUsingGlobalIndexSafely(predicate, mapContainer,
                 migrationStamp, initialPartitions.size());
 
+        if (entries != null && !initialPartitions.equals(actualPartitions)) {
+            assert indexes.isGlobal();
+            // if the query runs on a subset of partitions, filter the results from a global index
+            entries = IterableUtil.filter(entries,
+                    e -> {
+                        int partitionId = HashUtil.hashToIndex(e.getKeyData().getPartitionHash(), partitionCount);
```

Comment:
That leads to the computation of the hash code for every scanned entry, basically a call to `HashUtil.MurmurHash3_x86_32(byte[], int, int)` which is pretty heavy. It doesn't look critical (especially for small keys) since this path should be triggered only in special cases, but having such an overhead in the hottest loop of the scan isn't good. Let's create an issue at least, WDYT?

Reply:
Followup issue created: #18325